File: mobile-ios/treasuryflow/UI/Examples/ExampleListViewModel.swift
import Foundation
import Observation

enum ExampleListViewState: Equatable {
    case idle
    case loading
    case success
    case error(String)
}

@MainActor
@Observable
final class ExampleListViewModel {
    var viewState: ExampleListViewState = .idle
    var examples: [ExampleResponse] = []
    var currentPage: Int = 0
    var totalPages: Int = 0
    var isLoadingMore: Bool = false

    private let exampleRepository: ExampleRepository
    private let authRepository: AuthRepository

    init(exampleRepository: ExampleRepository, authRepository: AuthRepository) {
        self.exampleRepository = exampleRepository
        self.authRepository = authRepository
    }

    private func shouldDeferToAuthGate(for error: Error) -> Bool {
        guard case APIError.unauthorized = error else { return false }
        viewState = .idle
        return true
    }

    func loadExamples() async {
        viewState = .loading
        do {
            let response = try await exampleRepository.list(page: 0, size: 20)
            examples = response.content
            currentPage = response.page
            totalPages = response.totalPages
            viewState = .success
        } catch {
            if shouldDeferToAuthGate(for: error) { return }
            viewState = .error(error.localizedDescription)
        }
    }

    func loadMore() async {
        guard !isLoadingMore, currentPage + 1 < totalPages else { return }
        isLoadingMore = true
        defer { isLoadingMore = false }
        do {
            let response = try await exampleRepository.list(page: currentPage + 1, size: 20)
            examples.append(contentsOf: response.content)
            currentPage = response.page
            totalPages = response.totalPages
        } catch {
            if shouldDeferToAuthGate(for: error) { return }
            // Silently fail on load more
        }
    }

    func deleteExample(_ id: String) async {
        do {
            try await exampleRepository.delete(id)
            examples.removeAll { $0.id == id }
        } catch {
            if shouldDeferToAuthGate(for: error) { return }
            // Show error via snackbar if needed
        }
    }

    func logout() async {
        await authRepository.logout()
        viewState = .idle
    }
}

File: mobile-android/app/src/main/kotlin/com/mortitech/treasuryflow/feature/example/ui/ExampleListViewModel.kt
package com.mortitech.treasuryflow.feature.example.ui

import com.mortitech.treasuryflow.R
import com.mortitech.treasuryflow.core.common.AppResult
import com.mortitech.treasuryflow.designsystem.text.UiText
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mortitech.treasuryflow.feature.auth.data.repository.AuthRepository
import com.mortitech.treasuryflow.feature.example.data.repository.ExampleRepository
import com.mortitech.treasuryflow.feature.example.domain.model.Example
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

sealed interface ExampleListUiState {
    data object Idle : ExampleListUiState
    data object Loading : ExampleListUiState
    data class Success(
        val examples: List<Example>,
        val page: Int,
        val totalPages: Int,
        val isLoadingMore: Boolean = false,
    ) : ExampleListUiState
    data class Error(val message: UiText) : ExampleListUiState
}

@HiltViewModel
class ExampleListViewModel @Inject constructor(
    private val exampleRepository: ExampleRepository,
    private val authRepository: AuthRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow<ExampleListUiState>(ExampleListUiState.Idle)
    val uiState: StateFlow<ExampleListUiState> = _uiState.asStateFlow()

    init {
        loadExamples()
    }

    fun loadExamples() {
        viewModelScope.launch {
            _uiState.value = ExampleListUiState.Loading
            when (val result = exampleRepository.list(page = 0)) {
                is AppResult.Success -> {
                    _uiState.value = ExampleListUiState.Success(
                        examples = result.data.content,
                        page = result.data.page,
                        totalPages = result.data.totalPages,
                    )
                }
                is AppResult.Error -> {
                    _uiState.value = ExampleListUiState.Error(
                        result.message.toUiText(R.string.error_failed_load_examples),
                    )
                }
            }
        }
    }

    fun loadMore() {
        val current = _uiState.value as? ExampleListUiState.Success ?: return
        if (current.isLoadingMore || !current.hasNextPage) return

        _uiState.value = current.copy(isLoadingMore = true)

        viewModelScope.launch {
            when (val result = exampleRepository.list(page = current.page + 1)) {
                is AppResult.Success -> {
                    _uiState.value = ExampleListUiState.Success(
                        examples = current.examples + result.data.content,
                        page = result.data.page,
                        totalPages = result.data.totalPages,
                    )
                }
                is AppResult.Error -> {
                    _uiState.value = current.copy(isLoadingMore = false)
                }
            }
        }
    }

    fun deleteExample(id: String) {
        val current = _uiState.value as? ExampleListUiState.Success ?: return
        _uiState.value = current.copy(
            examples = current.examples.filter { it.id != id },
        )
        viewModelScope.launch {
            when (exampleRepository.delete(id)) {
                is AppResult.Success -> Unit
                is AppResult.Error -> loadExamples()
            }
        }
    }

    fun logout() {
        viewModelScope.launch {
            authRepository.logout()
        }
    }

    private val ExampleListUiState.Success.hasNextPage: Boolean
        get() = page < totalPages - 1

    private fun String.toUiText(fallbackResId: Int): UiText =
        if (isNotBlank()) {
            UiText.DynamicString(this)
        } else {
            UiText.StringResource(fallbackResId)
        }
}
